alf/utils/multi_gpu_utils_test.py

```diff
@@ -0,0 +1,129 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+import warnings
+import numpy as np
+
+import torch
+import torch.nn as nn
+from torch.utils.data import Dataset, DataLoader
+
+from alf.utils.multi_gpu_utils import MultiGPU
+
+
+class Model(nn.Module):
+    def __init__(self, input_size, output_size):
+        super(Model, self).__init__()
+        self.fc = nn.Linear(input_size, output_size)
+
+    def forward(self, input):
+        # the standard interface
+        output = self.fc(input)
+        global in_model_batch_size
+        in_model_batch_size = input.size(0)
+        return output
+
+    def train_step(self, input):
+        # self-defined non-standard interface
+        output = self.fc(input)
+        return output
+
+
+class DummyDataset(Dataset):
+    def __init__(self, size, length):
+        self.len = length
+        self.data = torch.randn(length, size)
+
+    def __getitem__(self, index):
+        return self.data[index]
+
+    def __len__(self):
+        return self.len
+
+
+# some shared parameters for testing
+input_size = 4
+output_size = 3
+
+batch_size = 30
+data_size = 50
+
+in_model_batch_size = 0
+
+
+class TestSingleGPU(unittest.TestCase):
+    def test_single_gpu(self):
+        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        model = Model(input_size, output_size).to(device)
+
+        gpu_ids = [0]
+        model = MultiGPU(model, gpu_ids)
+
+        data_loader = DataLoader(
+            dataset=DummyDataset(input_size, data_size),
+            batch_size=batch_size,
+            shuffle=True)
+
+        for data in data_loader:
+            input = data.to(device)
+            output = model(input)
+            outside_batch_size = input.size(0)
+
+            self.assertTrue(in_model_batch_size * len(gpu_ids),
+                            outside_batch_size)
+
+
+class TestMultiGPU(unittest.TestCase):
+    def test_multi_gpu(self):
+        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        model = Model(input_size, output_size).to(device)
+
+        gpu_ids = [0, 1]
+        model = MultiGPU(model, gpu_ids)
+
+        data_loader = DataLoader(
+            dataset=DummyDataset(input_size, data_size),
+            batch_size=batch_size,
+            shuffle=True)
+
+        for data in data_loader:
+            input = data.to(device)
+            output = model(input)
+            outside_batch_size = input.size(0)
+
+            self.assertTrue(in_model_batch_size * len(gpu_ids),
+                            outside_batch_size)
+
+
+class TestMultiGPUCustomizedFunction(unittest.TestCase):
+    def test_multi_gpu_customized_function(self):
+        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        model = Model(input_size, output_size).to(device)
```

Comment:
Do we always have to first put model on one of the gpus?

Reply:
yes